readthedocs/projects/tasks/builds.py

```diff
@@ -0,0 +1,1020 @@
+    before_vcs,
+    files_changed,
+)
+
+from .mixins import SyncRepositoryMixin
+from .utils import clean_build, BuildRequest, send_external_build_status
+from .search import fileify
+
+log = structlog.get_logger(__name__)
+
+
+class SyncRepositoryTask(SyncRepositoryMixin, Task):
+
+    """
+    Entry point to synchronize the VCS documentation.
+
+    This task checks all the branches/tags from the external repository (by
+    cloning) and update/sync the versions (by hitting the API) we have
+    stored in the database to match these branches/tags.
+
+    This task is executed on the builders and use the API to update the version
+    in our database.
+    """
+
+    name = __name__ + '.sync_repository_task'
+    max_retries = 5
+    default_retry_delay = 7 * 60
+
+    def before_start(self, task_id, args, kwargs):
+        log.info('Running task.', name=self.name)
+
+        # NOTE: save all the attributes to do a clean up when finish
+        self._attributes = list(self.__dict__.keys()) + ['_attributes']
+
+        self.environment_class = DockerBuildEnvironment
+        if not settings.DOCKER_ENABLE:
+            # TODO: delete LocalBuildEnvironment since it's not supported
+            # anymore and we are not using it
+            self.environment_class = LocalBuildEnvironment
+
+        version_id = kwargs.get('version_id')
+
+        # load all data from the API required for the build
+        self.version = self.get_version(version_id)
+        self.project = self.version.project
+
+        log.bind(
+            project_slug=self.project.slug,
+            version_slug=self.version.slug,
+        )
+
+        self.project.repo_nonblockinglock(version=self.version).acquire_lock()
+
+    def on_failure(self, exc, task_id, args, kwargs, einfo):
+        # Do not log as error handled exceptions
+        if isinstance(exc, RepositoryError):
+            log.warning(
+                'There was an error with the repository.',
+                exc_info=True,
+            )
+        elif isinstance(exc, LockTimeout):
+            countdown = 120
+            log.info(
+                'Lock still active. Retrying this task with countdown delay...',
+                countdown=countdown,
+            )
+            self.retry(
+                exc=exc,
+                throw=False,
+                countdown=countdown,
+            )
+        else:
+            # Catch unhandled errors when syncing
+            log.exception('An unhandled exception was raised during VCS syncing.')
+
+    def after_return(self, status, retval, task_id, args, kwargs, einfo):
+        clean_build(self.version)
+
+        # unlock repository directory
+        self.project.repo_nonblockinglock(version=self.version).release_lock()
+
+        # HACK: cleanup all the attributes set by the task under `self`
+        for attribute in list(self.__dict__.keys()):
+            if attribute not in self._attributes:
+                del self.__dict__[attribute]
```

Comment:
"Yes" to everything you noted here.

I ended up with this ugly hack for now because I didn't find a way to be able to use all the Celery machinery of the handlers by using `TaskStep` idea. I'm happy to get some help on this if we can find a better way to solve this problem.

It may not be hard, but what I've tried didn't work at first and I wanted to continue moving forward instead of being block immediately when I started the refactor 😊 

Also, note that we are not executing tasks in parallel now. Our builders only run 1 concurrent build task.

Reply:
Notes from meeting on a path forward perhaps worth exploring here:

* Move our attributes from `self` to a namespaced `self.data = {}`. Don't manipulate attributes at `self` on `Task`
* Or, consider usage of the TaskStep pattern to encapsulate data. This probably still needs a namespace and dictionary lookup
* Maybe use `task_id` or build pk from args for `self.data` key lookup
* We might need this to be a cache-like object with TTL or it could produce a memory leak as builds accumulate. Deleting keys from the data namespace should be fairly accurate however.

```
task_data = self.data[task_id]
# All of the attributes on `self` on `Task` are in `self.data[x]` instead
build = task_data.build
```

Reply:
I created a "namespace" to store all the task-related data under `self.data` and do a cleanup each time a new task is executed. This way, we will avoid the problem of shared data between different execution tasks.

Please, take a look at it at https://github.com/readthedocs/readthedocs.org/pull/8874 (I created a different PR to make it easier to review, but basically is a search&replace of `self.` by `self.data.`)

Let me know if the new pattern is aligned with the ideas that we discussed in the meeting.

Reply:
Yeah, this was the first step that we talked about, and I'm happy to have it included in this initial batch of work.

We do need to think of ways to make this thread safe however, but we can come back to that after merging this. I don't think we were too concerned about complicating the implementation here for this use case.